### PR TITLE
Add DateTime Leap Seconds Tests

### PR DESCRIPTION
**Pull request by tarekgh**
https://github.com/dotnet/coreclr/pull/21420

### CODE REVIEW

File: src/System.Runtime/tests/System/DateTimeTests.cs [hunk 1]
```diff
@@ -2177,6 +2178,59 @@ public void GetObjectData_NullInfo_ThrowsArgumentNullException()
             AssertExtensions.Throws<ArgumentNullException>("info", () => ((ISerializable)DateTime.Now).GetObjectData(null, new StreamingContext()));
         }
 
+        [Fact]
+        public void TestRoundTrippingDateTimeAndFileTime()
+        {
+            // This test ensure the round tripping of DateTime with the system file time.
+            // It is important to have this working on systems support leap seconds as the conversion wouldn't be simple
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DateTimeTests.cs]: Nit: on systems supporting leap seconds

File: src/System.Runtime/tests/System/DateTimeTests.cs [hunk 2]
```diff
@@ -2177,6 +2178,59 @@ public void GetObjectData_NullInfo_ThrowsArgumentNullException()
...
+            DateTime now = DateTime.UtcNow;
+            long fileTime = now.ToFileTimeUtc();
+            DateTime roundTrippedDateTime = DateTime.FromFileTimeUtc(fileTime);
+            Assert.Equal(now, roundTrippedDateTime);
+
+            now = DateTime.Now;
+            fileTime = now.ToFileTime();
+            roundTrippedDateTime = DateTime.FromFileTime(fileTime);
+            Assert.Equal(now, roundTrippedDateTime);
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DateTimeTests.cs]: Do we need a similar test for DateTimeOffset?

File: src/System.Runtime/tests/System/DateTimeTests.cs [hunk 3]
```diff
@@ -2177,6 +2178,59 @@ public void GetObjectData_NullInfo_ThrowsArgumentNullException()
...
+        public void TestTimeSynchronizationWithTheSystem()
+        {
+            // The reported time by the framework should be synchronized with the OS.
+            // There shouldn't be any shift by more than one second, otherwise there is something wrong.
+            // This test is useful when running on a system supporting leap seconds to ensure when the system
+            // has leap seconds, the framework reported time will still be synchronized.
+
+            SYSTEMTIME st;
+            GetSystemTime(out st);
+            DateTime dt = DateTime.UtcNow;
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DateTimeTests.cs]: The test assumes these two calls will occur within a second from each other.  Any concern about that being flaky in CI on a heavily loaded system?  Worth a retry?

File: src/System.Runtime/tests/System/DateTimeTests.cs [hunk 4]
```diff
@@ -2177,6 +2178,59 @@ public void GetObjectData_NullInfo_ThrowsArgumentNullException()
...
+            internal ushort wYear;
+            internal ushort wMonth;
+            internal ushort wDayOfWeek;
+            internal ushort wDay;
+            internal ushort wHour;
+            internal ushort wMinute;
+            internal ushort wSecond;
+            internal ushort wMillisecond;
+        }
+
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DateTimeTests.cs]: Do we have or need tests that constructs a specific leap second time and validates that the right things happen?

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: DateTimeOffset is using DateTime for such operations.

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 3 above)
Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: I think it is worth to try.

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: I don't think we can do that in the current time as we don't have a system with a leap seconds.

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 3 above)
Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: If we see the test flaky in CI, we can do some workaround at that time. let's try it first and look what we'll get.

Comment (jkotas) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: Similar patterns have *always* proved to be flaky in the CI. I am pretty sure this is going to be flaky as written.

Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: Ok, I'll try to make it more reliable then.

**Comment by tarekgh** [comment]
@stephentoub @jkotas let me know if the last commit addressed your concern. thanks.

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 3 above)
Comment (danmosemsft) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: CI timing is variable  enough and so many tests run every day that tests need to be as close to bullet proof as possible.

File: src/System.Runtime/tests/System/DateTimeTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/System.Runtime/tests/System/DateTimeTests.cs]: We already have similar test for DateTimeOffset https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/DateTimeOffsetTests.cs#L518

**Comment by tarekgh** [comment]
merging this one as I am not sure I'll be able to do more follow up in current time. if there is any pending concerns we can track it with an issue if needed. thanks all.

**Pull request merged**